src/assign.c

```diff
@@ -134,22 +134,27 @@ static int _selfrefok(SEXP x, Rboolean checkNames, Rboolean verbose) {
     return 0;                      // # nocov ; see http://stackoverflow.com/questions/15342227/getting-a-random-internal-selfref-error-in-data-table-for-r
   if (x!=R_ExternalPtrAddr(prot) && !ALTREP(x))
     SET_TRUELENGTH(x, LENGTH(x));  // R copied this vector not data.table, it's not actually over-allocated
-  return checkNames ? names==tag : x==R_ExternalPtrAddr(prot);
+  return names==tag && x==R_ExternalPtrAddr(prot);
 }
 
-static Rboolean selfrefok(SEXP x, Rboolean verbose) {   // for readability
-  return(_selfrefok(x, FALSE, verbose)==1);
+static bool selfrefok(SEXP x, bool verbose) {
+  return _selfrefok(x, verbose)==1;
 }
-static Rboolean selfrefnamesok(SEXP x, Rboolean verbose) {
-  return(_selfrefok(x, TRUE, verbose)==1);
+
+SEXP selfrefokwrapper(SEXP x, SEXP verbose) {
+  if (!IS_TRUE_OR_FALSE(verbose))
+    error(_("%s must be TRUE or FALSE"), "verbose"); // # nocov
+  return ScalarInteger(_selfrefok(x, LOGICAL(verbose)[0]));
 }
 
-static SEXP shallow(SEXP dt, SEXP cols, R_len_t n)
+static SEXP _shallow(SEXP dt, SEXP cols, const int nSpare)
 {
-  // NEW: cols argument to specify the columns to shallow copy on. If NULL, all columns.
-  // called from alloccol where n is checked carefully, or from shallow() at R level
-  // where n is set to truelength (i.e. a shallow copy only with no size change)
+  // cols: the columns to shallow copy on; all if NULL
+  // nSpare: passed by user to setalloccol()/alloc.col(); otherwise <0 means the datatable.alloccol option 
+  if (TYPEOF(dt)!=VECSXP) error(_("dt passed to shallow isn't type VECSXP"));
+  if (!isNull(cols) && !isInteger(cols)) error(_("cols passed to shallow isn't NULL or integer"));
   int protecti=0;
+  const int n = (isNull(cols) ? length(dt) : length(cols)) + (nSpare<0 ? allocColOpt() : nSpare);
```

Comment:
Could move `const int l = isNull(cols) ? LENGTH(dt) : length(cols);` up from line from line 180 and already use it here.

Reply:
Thanks Ben. I have to admit I struggled in this PR so it's great to have more eyes on it helping out. Will come back to this line.